File: project_euler/02/023.py
#!/usr/bin/env python3
'''Copyright [2017] <dmnsn7@gmail.com>'''

N = 28123


def get_fac_sum(num_n):
    fac_sum, tmp_n = 1, num_n
    for i in range(2, num_n):
        if i > num_n // i:
            break
        if num_n % i == 0:
            to_multi, base = 1, 1
            while num_n % i == 0:
                base *= i
                to_multi += base
                num_n //= i

            fac_sum *= to_multi

    if num_n != 1:
        fac_sum *= num_n + 1

    return fac_sum - tmp_n


def is_abund(num_n):
    return get_fac_sum(num_n) > num_n


def main():
    abund = set()
    for i in range(1, N + 1):
        if is_abund(i):
            abund.add(i)

    add_up = 0
    for i in range(1, N + 1):
        flag = True
        for num in abund:
            if (num > i - num) or not flag:
                break
            if i - num in abund:
                flag = False
        add_up += i if flag else 0

    print(add_up)


if __name__ == '__main__':
    main()

File: project_euler/02/023.go
// Copyright [2017] <dmnsn7@gmail.com>

package main

import "fmt"

const cn = 28123

func getFacSum(n int) int {
	facSum, tn := 1, n
	for i := 2; i <= n/i; i++ {
		if n%i == 0 {
			toMulti, base := 1, 1
			for n%i == 0 {
				base *= i
				toMulti += base
				n /= i
			}

			facSum *= toMulti
		}
	}

	if n != 1 {
		facSum *= n + 1
	}

	return facSum - tn
}

func isAbund(n int) bool {
	return getFacSum(n) > n
}

func main() {
	mAbund := map[int]int{}
	vAbund := []int{}
	for i := 1; i <= cn; i++ {
		if isAbund(i) {
			mAbund[i] = 0
			vAbund = append(vAbund, i)
		}
	}

	sum := 0
	for i := 1; i <= cn; i++ {
		flag := true
		for j := 0; j < len(vAbund) && vAbund[j] <= i-vAbund[j] && flag; j++ {
			if _, exist := mAbund[i-vAbund[j]]; exist {
				flag = false
			}
		}

		if flag {
			sum += i
		}
	}

	fmt.Printf("%d\n", sum)
}
